opm/simulators/utils/MPISerializer.hpp

```diff
@@ -0,0 +1,373 @@
+
+template<class T, class H, class KE, class A>
+void unpack(std::unordered_set<T,H,KE,A>& data,
+            std::vector<char>& buffer, int& position,
+            Opm::Parallel::MPIComm comm);
+
+template<class T, size_t N>
+void unpack(std::array<T,N>& data, std::vector<char>& buffer, int& position,
+          Opm::Parallel::MPIComm comm);
+
+template<class T1, class T2, class C, class A>
+void unpack(std::map<T1,T2,C,A>& data, std::vector<char>& buffer, int& position,
+            Opm::Parallel::MPIComm comm);
+
+template<class T1, class T2, class H, class P, class A>
+void unpack(std::unordered_map<T1,T2,H,P,A>& data, std::vector<char>& buffer, int& position,
+            Opm::Parallel::MPIComm comm);
+
+void unpack(char* str, std::size_t length, std::vector<char>& buffer, int& position,
+            Opm::Parallel::MPIComm comm);
+
+template<std::size_t Size>
+void unpack(std::bitset<Size>& data, std::vector<char>& buffer, int& position,
+            Opm::Parallel::MPIComm comm);
+
+/// prototypes for complex types
+
+#define ADD_PACK_PROTOTYPES(T) \
+  std::size_t packSize(const T& data, Opm::Parallel::MPIComm comm); \
+  void pack(const T& data, std::vector<char>& buffer, int& position, \
+          Opm::Parallel::MPIComm comm); \
+  void unpack(T& data, std::vector<char>& buffer, int& position, \
+              Opm::Parallel::MPIComm comm);
+
+ADD_PACK_PROTOTYPES(std::string)
+ADD_PACK_PROTOTYPES(time_point)
+
+template<typename T, typename... Args>
+void variadic_packsize(size_t& size, Parallel::Communication comm, T& first, Args&&... args)
+{
+  size += packSize(first, comm);
+  if constexpr (sizeof...(args) > 0)
+      variadic_packsize(size, comm, std::forward<Args>(args)...);
+}
+
+template<typename T, typename... Args>
+void variadic_pack(int& pos, std::vector<char>& buffer, Parallel::Communication comm, T& first, Args&&... args)
+{
+  pack(first, buffer, pos, comm);
+  if constexpr (sizeof...(args) > 0)
+      variadic_pack(pos, buffer, comm, std::forward<Args>(args)...);
+}
+
+template<typename T, typename... Args>
+void variadic_unpack(int& pos, std::vector<char>& buffer, Parallel::Communication comm, T& first, Args&&... args)
+{
+  unpack(first, buffer, pos, comm);
+  if constexpr (sizeof...(args) > 0)
+      variadic_unpack(pos, buffer, comm, std::forward<Args>(args)...);
+}
+
+#if HAVE_MPI
+template<typename... Args>
+void broadcast(Parallel::Communication comm, int root, Args&&... args)
+{
+  if (comm.size() == 1)
+      return;
+
+  size_t size = 0;
+  if (comm.rank() == root)
+      variadic_packsize(size, comm, std::forward<Args>(args)...);
+
+  comm.broadcast(&size, 1, root);
+  std::vector<char> buffer(size);
+  if (comm.rank() == root) {
+      int pos = 0;
+      variadic_pack(pos, buffer, comm, std::forward<Args>(args)...);
+  }
+  comm.broadcast(buffer.data(), size, root);
+  if (comm.rank() != root) {
+      int pos = 0;
+      variadic_unpack(pos, buffer, comm, std::forward<Args>(args)...);
+  }
```

Comment:
Quick note here: We are generally not allowed to `forward<>` arguments more than once.  I am guilty of having done that myself in the past, but now I try hard to avoid it.  Forwarding probably works in this particular case since I expect the arguments to `broadcast()` to be l-values instead of pr- or x-values in which case the reference collapsing process on the forwarding references does nothing at all.

Reply:
You are probably correct. However, this is not new code, it has been working for a year or so. Not sure how how I can avoid it, as we need to go over the params twice, once to find the size and once to pack them.

Reply:
I can probably remove the forward, it was added upon your request, https://github.com/OPM/opm-simulators/pull/3674#discussion_r745524730 :)

Reply:
> I can probably remove the forward

Cool.

> it was added upon your request

Right.  For what it's worth I still stand by that comment and it applies to the implementations of the individual `variadic_*()` operations.  Just not to implementation of the `broadcast()` operation.  You **can** `forward<>()` the arguments to the **last** part of the operation (i.e., the `variadic_unpack()`) however.

Reply:
done.